app/controllers/idv/scan_id_controller.rb

```diff
@@ -0,0 +1,273 @@
+      'Name': '',
+      'StorePII': false,
+    }].freeze
+    CLASSIFICATION_DATA = {
+      'Type': {
+      },
+    }.freeze
+    GOOD_DOCUMENT = { 'Result': 1, 'Fields': [{}] }.freeze
+    USER_SESSION_FLOW_ID = 'idv/doc_auth_v2'.freeze
+
+    def new
+      SecureHeaders.append_content_security_policy_directives(request,
+                                                              script_src: ['\'unsafe-eval\''])
+      render layout: false
+    end
+
+    def scan_complete
+      if all_checks_passed?
+        token_user_id ? continue_to_ssn_on_desktop : continue_to_ssn
+      else
+        idv_failure
+      end
+      clear_scan_id_session
+    end
+
+    def subscriptions
+      render_json SUBSCRIPTION_DATA
+    end
+
+    def instance
+      session[:scan_id] = {}
+      render_json(proxy_request { assure_id.create_document })
+    end
+
+    def image
+      assure_id.instance_id = params[:instance_id]
+      result = assure_id.post_image(request.body.read, params[:side].to_i)
+      render_json result
+    end
+
+    def classification
+      # use service when we accept multiple document types and dynamically decide # of sides
+      render_json CLASSIFICATION_DATA
+    end
+
+    def document
+      instance_id = params[:instance_id]
+      assure_id.instance_id = instance_id
+      data = proxy_request { assure_id.document }
+      return unless data
+      render_json process_data_and_store_pii_in_session_if_document_passes(data, instance_id)
+    end
+
+    def field_image
+      render_json({})
+    end
+
+    def liveness
+      live_face_image = process_selfie
+      return unless live_face_image
+      scan_id_session[:liveness_pass] = true
+
+      face_image_from_document = fetch_face_from_document
+      return unless face_image_from_document
+
+      check_face_match(face_image_from_document, live_face_image)
+      render_json({})
+    end
+
+    def facematch
+      render_json({})
+    end
+
+    private
+
+    def check_face_match(image1, image2)
+      data = proxy_request { facematch_service.facematch(facematch_body(image1, image2)) }
+      return unless data
+      scan_id_session[:facematch_pass] = facematch_pass?(data)
+    end
+
+    def facematch_body(image1, image2)
+      { 'Data':
+        { 'ImageOne': image1,
+          'ImageTwo': image2 },
+        'Settings':
+          { 'SubscriptionId': Figaro.env.acuant_assure_id_subscription_id } }.to_json
+    end
+
+    def fetch_face_from_document
+      assure_id.instance_id = scan_id_session[:instance_id]
+      data = proxy_request { assure_id.face_image }
+      return if data.blank?
+      Base64.strict_encode64(data)
+    end
+
+    def process_selfie
+      liveness_body = request.body.read
+      liveness_data = proxy_request { liveness_service.liveness(liveness_body) }
+      return unless liveness_data
+      unless selfie_live?(liveness_data)
+        render_json({})
+        return
+      end
+      JSON.parse(liveness_body)['Image']
```

Comment:
this is "doing work" in a controller -- can we consider moving this logic to a Service or a Form object?

Reply:
I definitely considered moving some guts to a service or two.  First plan was to figure out what I actually needed the service to do.  After stream of conscious coding it a lot of stuff is tied to controller, the params, and the session.  So I'd be passing a lot of info and handling a lot of conditionals on the way back.  I'll give it a stab and see if it cleans it up somewhat.

Reply:
Definitely not a form object.  I think we shifted a ton of logic to form objects erroneously.

Reply:
service objects coming up...

Reply:
done